src/session-pool.js

```diff
@@ -377,151 +307,75 @@ SessionPool.prototype.requestStream = function(config) {
 };
 
 /**
- * Returns the current size of the pool.
- *
- * @returns {number}
- */
-SessionPool.prototype.size = function() {
-  return this.available() + this.borrowed();
-};
-
-/**
- * Attempts to borrow a session from the pool.
+ * Returns a read session
  *
- * @private
- *
- * @param {string} type The desired type to borrow.
- * @returns {Promise<Session>}
+ * @returns {Promise}
  */
-SessionPool.prototype.acquireSession_ = function(type) {
-  var self = this;
+SessionPool.prototype.getReadSession = function() {
+  const self = this;
 
-  if (!this.isOpen) {
+  if (!self.isOpen) {
     return Promise.reject(new Error('Database is closed.'));
   }
+  /*
+    if readpool is maxed and there are no available session
+    and if writePool has any available session then use it
+  */
+
+  if (
+    self.readPool.spareResourceCapacity === 0 &&
+    self.readPool.available === 0 &&
+    (self.writePool.spareResourceCapacity || self.writePool.available > 0)
+  ) {
+    return self.getWriteSession();
+  }
 
-  var trace = stackTrace.get();
-
-  return this.getSession_(type).then(function(session) {
-    session.lastUsed = Date.now();
-
-    self.traces_.set(session.id, trace);
-
-    if (!self.available()) {
-      self.emit('empty');
-    }
-
-    return session;
-  });
-};
-
-/**
- * Moves a session into the borrowed group.
- *
- * @private
- *
- * @param {Session} session The session object.
- */
-SessionPool.prototype.borrowSession_ = function(session) {
-  this.spliceSession_(session);
-  this.borrowed_.push(session);
-};
-
-/**
- * Creates a read session.
- *
- * @private
- *
- * @return {Promise<Session>}
- */
-SessionPool.prototype.createReadSession_ = function() {
-  var session = this.session_();
-
-  return session.create().then(function() {
-    session.type = READONLY;
-    return session;
-  });
-};
-
-/**
- * Attemps to create a write session - may return a read session.
- *
- * @private
- *
- * @return {Promise<Session>}
- */
-SessionPool.prototype.createWriteSession_ = function() {
-  var self = this;
-  var session = this.session_();
-
-  return session
-    .create()
-    .then(function() {
-      session.type = READWRITE;
-
-      return self.createTransaction_(session).catch(function() {
-        session.type = READONLY;
-      });
-    })
-    .then(function() {
-      return session;
-    });
+  return self.acquireQueue_.add(() => self.readPool.acquire());
```

Comment:
I think if we get this far we want to check that there are actually read sessions available before acquiring, if not then we want to watch for resources being freed up in either pool and acquire the first available. The same logic should be applied to `getWriteSession`.

Reply:
I have made these changes and will push my changes